timemachine/cpp/src/kernels/k_nonbonded.cuh

```diff
@@ -54,6 +54,38 @@ void __global__ k_check_rebuild_coords_and_box_gather(
     }
 }
 
+template <typename RealType, int COORDS_DIM, int PARAMS_DIM>
+void __global__ k_rescale_coords(
+    const int NUM_ROW_ATOMS,
+    const unsigned int *__restrict__ row_atom_idxs,
+    const unsigned int *__restrict__ row_anchor_idxs,
+    const double *__restrict__ coords,
+    const double *__restrict__ params,
+    double *__restrict__ coords_out) {
+    static_assert(COORDS_DIM == 3);
+    static_assert(PARAMS_DIM == 4);
+    int atom_idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if (atom_idx >= NUM_ROW_ATOMS) {
+        return;
+    }
+
+    unsigned int anchor_idx = row_anchor_idxs[atom_idx];
+
+    RealType anchor_x = coords[anchor_idx * COORDS_DIM + 0];
+    RealType anchor_y = coords[anchor_idx * COORDS_DIM + 1];
+    RealType anchor_z = coords[anchor_idx * COORDS_DIM + 2];
+    RealType rescale =
+        static_cast<RealType>(1.0) - static_cast<RealType>(params[atom_idx * PARAMS_DIM + 3]); // old w_coordinates
+
+    // Coords have 3 dimensions, params have 4
+#pragma unroll COORDS_DIM
+    for (int i = 0; i < COORDS_DIM; i++) {
+        RealType atom_pos = coords[atom_idx * COORDS_DIM + i];
+        RealType anchor_pos = coords[anchor_idx * COORDS_DIM + i];
+        coords_out[atom_idx * COORDS_DIM + i] = anchor_pos + rescale * (atom_pos - anchor_pos);
```

Comment:
nit: could this be simplified to

```suggestion
        coords_out[atom_idx * COORDS_DIM + i] = atom_pos + params[atom_idx * PARAMS_DIM + 3] * (anchor_pos - atom_pos);
```

(removing `rescale` definition)?

Reply:
rescale is removed by the compiler (would prefer to make the rescaling explicit).